decisions/0018-nmdc_client.md

```diff
@@ -0,0 +1,80 @@
+---
+# Update these YAML values so they describe this decision. Delete the leading `→` characters.
+status: proposed
+date: 2024-11-08
+deciders: Shreyas Cholia, Sierra Moxon, Jeff Baumes, Jing Cao, Mark Miller, Katherine Heal, Brynn 
+consulted: Sujay Patil, Alicia Clum  
+informed: Chris Mungall 
+---
+# Use of nmdc_client
+
+## Context
+
+We have a number of different solutions to make complex queries against the runtime API, especially involving graph traversal. Jeff Baumes recently proposed an nmdc_client library that can capture some of the complexities and hide them from users.  
+
+### Summary of Jeff Baumes GDoc
+This document collects bits of work that have been done in constructing a python client for NMDC. Many have been disparately working on this problem.
+
+https://api.microbiomedata.org/docs#/metadata/list_from_collection_nmdcschema__collection_name__get
+- An endpoint that lets you get at the NMDC objects from any collection
+- You specify the collection name
+- You specify the query as a Mongo query object
+- Joins are not supported
+
+https://github.com/polyneme/nmdc-client-api/blob/main/index.ipynb
+- Breaks up ids into groups because of URL limit
+- Traverses certain types of links to move between types
+
+https://github.com/microbiomedata/nmdc_notebooks
+- The nmdc_api.py file has some patterns that have been useful for getting data and associating between NMDC types
+- It also breaks up ids into groups because of the URL size limit
+- Also enables traversal if you know the related collections and slot names
+- Has some code for working with dataframes
+
+https://api.microbiomedata.org/docs#/find/find_data_objects_for_study_data_objects_study__study_id__get
+- An endpoint that performs a join to find objects of a certain type (DataObject) from another (Study)
+- It uses the alldocs collection (an aggregation of our source of truth collections) to make the join
+- During the retreat, Jing also implemented another prototype endpoint to go in the other direction, from DataObject to Biosample
+
+Graph database
+- Mark Miller spun up a graph database view of NMDC data to help query and inspect connections between NMDC objects
+
+https://github.com/microbiomedata/refscan
+- A library that draws out connections between NMDC classes and constructs a network
+- Utilizes the actual schema and SchemaView to inspect NMDC slot relationships without requiring any hard-coded knowledge of NMDC (other than the Database class)
+- nmdc-server (Data portal)
+- Has an endpoint used by the client which can link across collections, but this is not “client friendly” in general and was designed to be consumed by the data portal.
```

Comment:
I think the author meant for this bullet point to be a top-level heading, as opposed to a bullet point under refscan.

```suggestion

nmdc-server (Data portal)
- Has an endpoint used by the client which can link across collections, but this is not “client friendly” in general and was designed to be consumed by the data portal.
```